### DISCUSSION TITLE
Update ethr DID using a web3 provider (such as Metamask)

### DISCUSSION THREAD

**Discussion by jrhender** [decentralized-identity/veramo#692] [Q&A] [answered]
I am looking for an example of Veramo client which has been configured to use a "web3 provider" such as Metamask to control an `ethr` DID Document. If anyone is aware of an example of this being done, it would be greatly appreciated 🙇 

I think it would be possible to do using the `did-provider-ethr` package as the `txSigner` in `EthrDIDProvider` is a `KmsEthereumSigner` which calls `keyManagerSign` from the agent: https://github.com/uport-project/veramo/blob/4a5e0a288420111a947af023eec8b06cbc484e35/packages/did-provider-ethr/src/kms-eth-signer.ts#L39
So what would be required is a plugin implementing the `keyManagerSign` method which delegates to a web3 provider. Is this correct?

**Comment by mirceanis** [comment, marked as answer]
Yes, that is correct.
However, you might run into other issues if you replace the whole keyManager plugin.
Instead it would be better to create an implementation of an `AbstractKeyManagementSystem`(KMS) (take a look at `@veramo/kms-local`) that wraps the signing capability of a web3 provider.

We have some plans to create a KMS that does exactly that, but haven't yet had bandwidth to do the actual implementation.
You are welcome to give it a shot, of course.
Contributions are always welcome.

**Reply by mirceanis** [reply]
The existing `@veramo/key-manager` plugin can work with multiple KMS implementations at once.
In our examples and tests the `@veramo/kms-local` implementation is used, usually named `"local"` wherever a `kms` parameter is required.

**Comment by jrhender** [comment]
I think this may be related to https://github.com/uport-project/veramo/issues/688

**Reply by mirceanis** [reply]
Exactly!
